Lib/test/test_typing.py

```diff
@@ -894,6 +906,93 @@ class C(Generic[Unpack[Ts]]): pass
         self.assertNotEqual(C[Unpack[Ts1]], C[Unpack[Ts2]])
 
 
+class TypeVarTuplePicklingTests(BaseTestCase):
+    # These are slightly awkward tests to run, because TypeVarTuples (and things
+    # made generic using TypeVarTuples) are only picklable if defined in the
+    # global scope. We therefore need to push various things defined in these
+    # tests into the global scope with `global` statements at the start of each
+    # test.
+
+    @all_pickle_protocols
+    def test_pickling_then_unpickling_results_in_equality(self, proto):
+        global Ts1  # See explanation at start of class.
+        Ts1 = TypeVarTuple('Ts1')
+        Ts2 = pickle.loads(pickle.dumps(Ts1, proto))
+        self.assertEqual(Ts1, Ts2)
+
+    @all_pickle_protocols
+    def test_pickling_then_unpickling_unpacked_results_in_equality(self, proto):
+        global Ts  # See explanation at start of class.
+        Ts = TypeVarTuple('Ts')
+        unpacked1 = Unpack[Ts]
+        unpacked2 = pickle.loads(pickle.dumps(unpacked1, proto))
+        self.assertEqual(unpacked1, unpacked2)
+
+    @all_pickle_protocols
+    def test_pickling_then_unpickling_variadic_class_results_in_equality(
+            self, proto
+    ):
+        global T, Ts, A1, B1  # See explanation at start of class.
+        T = TypeVar('T')
+        Ts = TypeVarTuple('Ts')
+
+        class A1(Generic[Unpack[Ts]]): pass
+        A2 = pickle.loads(pickle.dumps(A1, proto))
+        self.assertEqual(A1, A2)
+
+        class B1(Generic[T, Unpack[Ts]]): pass
+        B2 = pickle.loads(pickle.dumps(B1, proto))
+        self.assertEqual(B1, B2)
+
+    @all_pickle_protocols
+    def test_pickling_then_unpickling_tuple_with_typevartuple_equality(
+            self, proto
+    ):
+        global T, Ts  # See explanation at start of class.
+        T = TypeVar('T')
+        Ts = TypeVarTuple('Ts')
+
+        a1 = Tuple[Unpack[Ts]]
+        a2 = pickle.loads(pickle.dumps(a1, proto))
+        self.assertEqual(a1, a2)
+
+        a1 = Tuple[T, Unpack[Ts]]
+        a2 = pickle.loads(pickle.dumps(a1, proto))
+        self.assertEqual(a1, a2)
+
+        a1 = Tuple[int, Unpack[Ts]]
+        a2 = pickle.loads(pickle.dumps(a1, proto))
+        self.assertEqual(a1, a2)
+
+    @all_pickle_protocols
+    def test_pickling_then_unpickling_variadic_func_results_in_equality(
+            self, proto
+    ):
+        global T, Ts, a1, b1, c1, d1, e1  # See explanation at start of class.
+        T = TypeVar('T')
+        Ts = TypeVarTuple('Ts')
+
+        def a1(*args: Unpack[Ts]): pass
```

Comment:
As with classes, functions are pickled by name, so this test doesn't really test anything about TypeVarTuples.

Reply:
Done.